webknossos/webknossos/skeleton/entities.py

```diff
@@ -0,0 +1,479 @@
+                (graph.get_max_node_id() for graph in self.flattened_graphs()),
+                [0],
+            )
+        )
+
+    def flattened_graphs(self) -> Generator["WkGraph", None, None]:
+        for child in self.children:
+            if isinstance(child, Group):
+                yield from child.flattened_graphs()
+            else:
+                yield child
+
+    def flattened_groups(self) -> Generator["Group", None, None]:
+        for child in self.children:
+            if isinstance(child, Group):
+                yield child
+                yield from child.flattened_groups()
+
+    def as_legacy_group(self) -> "LegacyGroup":  # type: ignore
+
+        return legacy_wknml.Group(
+            self.id,
+            self.name,
+            children=[
+                g.as_legacy_group() for g in self.children if isinstance(g, Group)
+            ],
+        )
+
+    def __hash__(self) -> int:
+        return hash((self._nml.id, self.id))
+
+
+@attr.define()
+class Node:
+    position: Vector3
+    _nml: "NML"
+    id: int = attr.ib(init=False)
+    comment: Optional[str] = None
+    radius: Optional[float] = None
+    rotation: Optional[Vector3] = None
+    inVp: Optional[int] = None
+    inMag: Optional[int] = None
+    bitDepth: Optional[int] = None
+    interpolation: Optional[bool] = None
+    time: Optional[int] = None
+
+    is_branchpoint: bool = False
+    branchpoint_time: Optional[int] = None
+    _enforce_id: Optional[int] = None
+
+    def __attrs_post_init__(self) -> None:
+        if self._enforce_id is not None:
+            self.id = self._enforce_id
+        else:
+            self.id = self._nml.element_id_generator.__next__()
+
+        self.position = vector3_as_float(self.position)
+
+    def __hash__(self) -> int:
+        return hash((self._nml.id, self.id))
+
+    def __eq__(self, other: object) -> bool:
+        return hash(self) == hash(other)
+
+
+@attr.define()
+class WkGraph:
+    """
+    Contains a collection of nodes and edges.
+    """
+
+    name: str
+    _nml: "NML"
+    color: Optional[Vector4] = None
+    id: int = attr.ib(init=False)
+    nx_graph: nx.Graph = attr.ib(init=False)
+    group_id: Optional[int] = None
+
+    _enforce_id: Optional[int] = None
+
+    def __attrs_post_init__(self) -> None:
+        self.nx_graph = nx.Graph()
+        if self._enforce_id is not None:
+            self.id = self._enforce_id
+        else:
+            self.id = self._nml.element_id_generator.__next__()
+
+    # todo: how to type this?
+    def get_nodes(self) -> Any:
+
+        return self.nx_graph.nodes
+
+    def get_node_positions(self) -> np.ndarray:
+        return np.array([node.position for node in self.nx_graph.nodes])
+
+    def get_node_by_id(self, node_id: int) -> Node:
+
+        # Todo: Use hashed access
+        for node in self.get_nodes():
+            if node.id == node_id:
+                return node
+        raise ValueError(f"No node with id {node_id} was found")
```

Comment:
If we simply subclass `nx.Graph`, we might also want to enable lookups by id like this:
```python
node = g.nodes[3]
```

I think this might be possible using some magic, and I think it would have a better UX. What do you think about this?

Reply:
Let's leave it as-is for now, we might look into nx-graph subclassing again later.